### PR TITLE
Fix Empty IdealState Calculation with Disabled Nodes in AutoRebalance…

### PR DESCRIPTION
**Pull request by MarkGaox**
…Strategy

### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:
- #2876 

(#200 - Link your issue number here: You can write "Fixes #XXX". Please use the proper keyword so that the issue gets closed automatically. See https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue
Any of the following keywords can be used: close, closes, closed, fix, fixes, fixed, resolve, resolves, resolved)

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:
(Write a concise description including what, why, how)

When calculating the preferred assignment, we should skip the node if the node is already disabled. 
* made above in 
 ```       
private Map<Replica, Node> computeExistingPreferredPlacement(
      final Map<String, Map<String, String>> currentMapping) {
```
* refine the naming in AutoRebalanceStrategy to follow the naming conventions.
### Tests

- [X] The following tests are written for this issue:

` mvn clean install -Dmaven.test.skip.exec=true && mvn test -o -Dtest=TestAutoRebalanceStrategy -pl=helix-core`

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Review by xyuanlu — COMMENTED** [review]
May I suggest separate the logic change and style change? It is easier to review and track change history.

**Review by junkaixue — COMMENTED** [review]
Split the PR with fix + refactor.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AutoRebalanceStrategy.java [hunk 1]
```diff
@@ -560,7 +560,7 @@ private Map<Replica, Node> computeExistingPreferredPlacement(
           Replica replica = new Replica(partition, replicaId);
           if (_preferredAssignment.containsKey(replica)
               && !existingPreferredAssignment.containsKey(replica)
-              && _preferredAssignment.get(replica).id == node.id) {
+              && _preferredAssignment.get(replica).id == node.id && node.isAlive) {
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AutoRebalanceStrategy.java:563]: discussed offline. Need refine behavior consistent with current delay auto rebalanced. It means when an instance disabled, it should still show in preference list but in map field it showed as INITIAL state.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AutoRebalanceStrategy.java [hunk 2]
```diff
@@ -336,13 +347,21 @@ private void prepareResult(ZNRecord znRecord) {
...
+
+    String stateModelDef = clusterData.getIdealState(_resourceName).getStateModelDefRef();
+    StateModelDefinition stateModel = clusterData.getStateModelDef(stateModelDef);
 
     // generate preference maps based on the preference lists
     for (String partition : _partitions) {
       List<String> preferenceList = znRecord.getListField(partition);
       int i = 0;
       for (String participant : preferenceList) {
+        // if the participant is not alive, assign it to the initial state
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AutoRebalanceStrategy.java:360]: If the node is not alive, why we have in the map? I assume we fix disabled ones, right?

My assumption is if the node is not alive, we should not show something in map field as this is strategy, this is not delay auto.

Only delay auto consider offline instances with offline timeout and put it as offline when node is not alive.

**Review by junkaixue — APPROVED** [review]

File: helix-core/src/test/java/org/apache/helix/controller/rebalancer/TestAutoRebalanceStrategy.java [hunk 3]
```diff
@@ -866,4 +907,220 @@ public void testWontMoveSinglePartitionUnnecessarily() {
...
+      Map<String, String> idealStateMap = znRecord.getMapField(partition);
+      currentMapping.put(partition, idealStateMap);
+    }
+    currentMapping.get(PARTITIONS[0]).put(NODES[0], "ERROR");
+    // now disable node 0, and make sure the dataCache provides it. And add another node n10 to the
+    // cluster. We want to make sure the n10 can pick up another replica of partition 0,1,2.
+    allNodes = new ArrayList<>(allNodes);
+    liveNodes = new ArrayList<>(liveNodes);
+    liveNodes.remove(NODES[0]);
+    allNodes.add("n10");
```
Comment (zpinto) [inline helix-core/src/test/java/org/apache/helix/controller/rebalancer/TestAutoRebalanceStrategy.java:1083]: Before we add another node, can we also check that the we don't drop a partition on any of the remaining nodes after disabling the one with the error replica?

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AutoRebalanceStrategy.java [hunk 4]
```diff
@@ -127,15 +129,16 @@ public ZNRecord computePartitionAssignment(final List<String> allNodes, final Li
     _stateMap = generateStateMap();
 
     // compute the preferred mapping if all nodes were up
-    _preferredAssignment = computePreferredPlacement(sortedAllNodes);
+    _preferredAssignment = computePreferredPlacement(sortedAllNodes, clusterData);
 
     // logger.info("preferred mapping:"+ preferredAssignment);
```
[the remark this answers is not in the document]
Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AutoRebalanceStrategy.java:137]: the commented out log line

File: helix-core/src/test/java/org/apache/helix/controller/rebalancer/TestAutoRebalanceStrategy.java (hunk 3 above)
Comment (MarkGaox) [reply helix-core/src/test/java/org/apache/helix/controller/rebalancer/TestAutoRebalanceStrategy.java:1101]: TFTR, added that check.

**Comment by junkaixue** [comment]
let @zpinto take another look?

**Comment by junkaixue** [comment]
@MarkGaox can you follow up with Zac?

**Review by zpinto — APPROVED** [review]
LGTM

**Comment by MarkGaox** [comment]
This PR is ready to merge. Approved by @junkaixue @zpinto 

Commit message: Fix IdealState Calculation in AutoRebalanceStrategy for Disabled Nodes

**Pull request merged**